HPCC-10666 Operations on Files should not, by default, change the state of  the file. e.g. STD.File.Copy converts compressed files to uncompressed files.

AttilaVamos opened the pull request:
Add preservecompression parameter to preserve compression in copy

Extend filecompcopy.ecl to test preservecompression

Update filecompcopy.xml key file

Signed-off-by: Attila Vamos attila.vamos@gmail.com

AttilaVamos commented:
@richardkchapman @ghalliday please review.

AttilaVamos commented:
@richardkchapman @ghalliday please ignore it. There is some merge problem (clash with JSON related changes). I will resolve it and come back.

AttilaVamos commented:
@richardkchapman @ghalliday Merge problem fixed, please review.

ghalliday commented on dali/ft/daft.cpp:
```diff
@@ -57,6 +57,15 @@ void CDistributedFileSystem::copy(IDistributedFile * from, IDistributedFile * to
+    bool compressOutput = options->getPropBool("@compress");
+    bool preserveCompression = options->getPropBool("@preserveCompression");
+    LOG(MCdebugInfo, unknownJob, "DFS: compressInput:%d, compressOutput:%d, preserveCompression:%d ", compressInput, compressOutput, preserveCompression);
+    sprayer->setPreserveCompression(preserveCompression);
```
I don't think the value set by this call is ever used.

ghalliday commented on dali/ft/filecopy.cpp:
```diff
@@ -2973,7 +2985,10 @@ void FileSprayer::updateTargetProperties()
                      (stricmp(aname,"@owner")==0)||
-                     ((stricmp(aname,FArecordCount)==0)&&!gotrc))
+                     ((stricmp(aname,FArecordCount)==0)&&!gotrc) ||
+                     ((stricmp(aname,"@blockCompressed")==0)&&copyCompressed) ||
```
This looks suspiciously like some code that should be in the dafile code - copyAttribute(name, copyCompression); - otherwise when new attributes are added they may not be copied.  The whole copy should probably 

Probably a different pull request though.

ghalliday commented on dali/ft/filecopy.hpp:
```diff
@@ -71,7 +71,7 @@ class DALIFT_API FileFormat
 public:
     FileFormat(FileFormatType _type = FFTunknown, unsigned _recordSize = 0) : headerLength((unsigned)-1), footerLength((unsigned)-1), markup(FMTunknown)
-            { set(_type, _recordSize); maxRecordSize = 0; quotedTerminator = true;}
+            { set(_type, _recordSize); maxRecordSize = 0; quotedTerminator = true; preserveCompression = false;}
```
I don't think this flag is ever used, and doesn't really make sense to be here.

ghalliday commented:
A few bits of code that can be deleted, but otherwise it looks like the correct approach.

AttilaVamos commented:
@ghalliday Updated, please re-review.

richardkchapman commented on dali/dfu/dfurun.cpp:
```diff
@@ -1221,6 +1221,8 @@ class CDFUengine: public CInterface, implements IDFUengine
 
                         opttree->setPropBool("@quotedTerminator", options->getQuotedTerminator());
 
```
minor: why so many blank lines?

richardkchapman commented on dali/dfu/dfuwu.hpp:
```diff
@@ -172,6 +172,8 @@ interface IConstDFUoptions : extends IInterface
     virtual bool getRecordStructurePresent() const = 0;
 
     virtual bool getQuotedTerminator() const = 0;
+
```
minor: why so many blank lines?

richardkchapman commented on dali/dfuplus/main.cpp:
```diff
@@ -107,6 +107,7 @@ void handleSyntax()
     out.append("        diffkeysrc=<old-key-name>   -- use keydiff/keypatch (src old name)\n");
     out.append("        diffkeydst=<old-key-name>   -- use keydiff/keypatch (dst old name)\n");
     out.append("        multicopy=0|1   -- each destination part gets whole file\n");
+    out.append("        preservecompression=0|1 -- optional, default is 0 (don't preserve compression)\n");
```
Is 0 the best default?

AttilaVamos replied:
From Gavin: 
"If we are preserving backward compatibility, the cleanest way I can think of achieving that is to add a new boolean option to the spray function "preserveCompression" which defaults to false."

richardkchapman replied:
I'm not convinced that maintaining compatibility (i.e. retaining the erroneous behaviour by default) is more important here than making sure that normal usage of this function behaves as people would expect (so long as we are going into a major release - we an add a note to the redbook).

@ghalliday what do you think?

ghalliday replied:
The biggest problem would occur if they were despraying, and decompressing.

I hadn't spotted that it was only supported for copy.  If so then I think we probably could default it on, but we should check with tony kirk etc. if it is likely to cause a problem.

richardkchapman commented on esp/scm/ws_fs.ecm:
```diff
@@ -86,6 +86,8 @@ ESPStruct [nil_remove] DFUWorkunit
     [min_ver("1.09")] bool recordStructurePresent(false);
 
     [min_ver("1.10")] bool quotedTerminator(true);
+
```
minor: why so many blank lines?

richardkchapman commented on testing/regress/ecl/filecompcopy.ecl:
```diff
@@ -106,6 +107,21 @@ sequential (
+            true,                           // allowoverwrite
+            false,                          // replicate
+            ,                               // asSuperfile,
+            ,
```
Is this , needed if you are specifying the parameter by name?

ghalliday commented on dali/dfu/dfuwu.hpp:
```diff
@@ -248,6 +251,7 @@ interface IConstDFUfileSpec: extends IInterface
     virtual StringBuffer &getRawDirectory(StringBuffer &str) const = 0;
     virtual StringBuffer &getRawFileMask(StringBuffer &str) const = 0;
     virtual bool getRemoteGroupOverride() const = 0;
+    virtual bool getPreserveCompression() const = 0;
```
Is this ever used?  Isn't preservation an option of the copy rather than the file?

ghalliday commented on esp/services/ws_fs/ws_fsService.cpp:
```diff
@@ -387,6 +387,8 @@ static void DeepAssign(IEspContext &context, IConstDFUWorkUnit *src, IEspDFUWork
             if(version >=1.10)
                 dest.setQuotedTerminator(quotedTerminator);
+            if(version >=1.12)
+                dest.setPreserveCompression(file->getPreserveCompression());
```
Shouldn't this be src.getPreserveCompression()?

ghalliday commented on dali/ft/filecopy.cpp:
```diff
@@ -2500,6 +2500,16 @@ void FileSprayer::setSourceTarget(IFileDescriptor * fd, DaftReplicateMode mode)
+    compressOutput = compress;
+}
+
+void FileSprayer::setPreserveCompression(bool preserveComp)
```
I think this function is not called, and the member variable is not used.

ghalliday commented:
@AttilaVamos I forgot to assign back.  I agree with Richard's comment that copy should probably default on.

AttilaVamos commented:
@Gavin Updated, please re-review.

richardkchapman commented:
I think you meant to assign this back to @ghalliday

AttilaVamos commented:
Yes, I tried.

ghalliday commented on dali/dfu/dfuwu.cpp:
```diff
@@ -1680,6 +1680,18 @@ class CDFUfileSpec: public CLinkedDFUWUchild, implements IDFUfileSpec
         queryRoot()->setPropBool("@remoteGroupOverride",set);
     }
 
+    bool getPreserveCompression() const
```
Are these functions used?

AttilaVamos replied:
Yes, they used in runWu() (dfurun.cpp:1223) to pass parameter.

ghalliday replied:
The functions for the options are used, but I can't see any use for these 2 (IDFUFileSpec has not been modified)

ghalliday commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -963,6 +959,27 @@ FILESERVICES_API char * FILESERVICES_CALL fsfCopy(ICodeContext *ctx, const char
 
+FILESERVICES_API void FILESERVICES_CALL fsCopy(ICodeContext *ctx, const char * sourceLogicalName, const char *destinationGroup, const char * destinationLogicalName, const char * sourceDali, int timeOut, const char * espServerIpPort, int maxConnections, bool overwrite, bool replicate, bool asSuperfile, bool compress, bool forcePush, int transferBufferSize)
+{
+    CTXFREE(parentCtx, implementCopy(ctx, sourceLogicalName, destinationGroup, destinationLogicalName, sourceDali, timeOut, espServerIpPort, maxConnections, overwrite, replicate, asSuperfile, compress, forcePush, transferBufferSize, true));
```
Should this have, false as the last parameter - to ensure that old workunits don't change their functionality.

richardkchapman replied:
I think it's more important that new workunits work as expected - deliberately passing false is unlikely I think

ghalliday replied:
This function should never be called from a new workunit - it is only there for backward compatibility.
However....  I don't think it matters that much.

AttilaVamos replied:
Should I change or keep it?

richardkchapman replied:
Leave as is I think

richardkchapman commented:
Could do with squashing the commits and cleaning up the commit message as per guidelines

ghalliday commented:
@AttilaVamos meant to assign back to you

AttilaVamos commented:
Automated Smoketest Build: success

AttilaVamos committed:
HPCC-10666 Operations on Files should not, by default, change the state of
the file. e.g. STD.File.Copy converts compressed files to uncompressed files.

Add preservecompression parameter to preserve compression in copy

Extend filecompcopy.ecl to test preservecompression flag

Update filecompcopy.xml key file

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

AttilaVamos commented:
Automated Smoketest Build: success

AttilaVamos commented:
@richardkchapman updated, commit squashed. Please check and merge.
